File: bindings/rust/src/graph/batch.rs
//! Batch operations for Graph.

use crate::Result;
use super::Graph;

impl Graph {
    /// Batch upsert multiple nodes.
    pub fn upsert_nodes_batch<I, N, P, K, V, L>(&self, nodes: I) -> Result<()>
    where
        I: IntoIterator<Item = (N, P, L)>,
        N: AsRef<str>,
        P: IntoIterator<Item = (K, V)>,
        K: AsRef<str>,
        V: AsRef<str>,
        L: AsRef<str>,
    {
        for (node_id, props, label) in nodes {
            self.upsert_node(node_id.as_ref(), props, label.as_ref())?;
        }
        Ok(())
    }

    /// Batch upsert multiple edges.
    pub fn upsert_edges_batch<I, S, T, P, K, V, R>(&self, edges: I) -> Result<()>
    where
        I: IntoIterator<Item = (S, T, P, R)>,
        S: AsRef<str>,
        T: AsRef<str>,
        P: IntoIterator<Item = (K, V)>,
        K: AsRef<str>,
        V: AsRef<str>,
        R: AsRef<str>,
    {
        for (source, target, props, rel_type) in edges {
            self.upsert_edge(source.as_ref(), target.as_ref(), props, rel_type.as_ref())?;
        }
        Ok(())
    }
}

File: bindings/typescript/src/graph/batch.ts
// Batch upsert helpers — pure loops over upsertNode / upsertEdge.
//
// These take a Connection first so Graph can delegate in three lines (see
// graph/index.ts). Mirrors bindings/python/src/graphqlite/graph/batch.py.
//
// **No atomicity.** Each item is a separate upsert; if one throws partway
// through, the earlier items are already committed. These do NOT wrap the loop
// in a transaction — a deliberate reproduction of the Python behavior. For
// atomic bulk inserts, a future bulk API would be the place, not here.
import type { Connection } from '../connection.ts';
import { upsertNode } from './nodes.ts';
import { upsertEdge } from './edges.ts';

/** `[nodeId, props, label]` — one node for {@link upsertNodesBatch}. */
export type NodeBatchItem = [string, Record<string, unknown>, string];

/** `[sourceId, targetId, props, relType]` — one edge for {@link upsertEdgesBatch}. */
export type EdgeBatchItem = [string, string, Record<string, unknown>, string];

/**
 * Upsert many nodes by calling {@link upsertNode} once per item.
 *
 * Not atomic: a throw partway through leaves earlier nodes committed.
 */
export function upsertNodesBatch(conn: Connection, nodes: NodeBatchItem[]): void {
  for (const [nodeId, props, label] of nodes) {
    upsertNode(conn, nodeId, props, label);
  }
}

/**
 * Upsert many edges by calling {@link upsertEdge} once per item.
 *
 * Not atomic (see {@link upsertNodesBatch}). Note this does **not** pass an
 * `edgeId`, so a batch cannot create parallel edges between the same pair —
 * MERGE reuses the existing edge. This matches the Python binding.
 */
export function upsertEdgesBatch(conn: Connection, edges: EdgeBatchItem[]): void {
  for (const [sourceId, targetId, props, relType] of edges) {
    upsertEdge(conn, sourceId, targetId, props, relType);
  }
}
